File: WoofIOS/SwiftPackages/Presentation/Favorites/Tests/PresentationFavoritesTests/ViewModel/FavoritesViewModelTest.swift
import Combine
import DataSourceFakes
import Nimble
import PresentationFavorites
import Quick
import Repositories
import Swinject
import SwinjectAutoregistration
import TestUtils
import UtilsTestExtensions

final class FavoritesViewModelTest: QuickSpec {
    override class func spec() {
        var container: Container!
        var viewModel: FavoritesViewModel!
        var dogsRepository: DogsRepository!
        var favoritesRepository: FavoritesRepository!
        var cancellables = Set<AnyCancellable>()

        Given("I open the favorites tab with dogs loaded in the grid") {
            beforeEach {
                container = Container().injectForViewModelTests()
                TimeAdvancingFactory(container).withTestSchedulers()
                DogsDataSourceFactory(container).withDogs(count: 5)
                dogsRepository = container~>
                favoritesRepository = container~>
                viewModel = container~>
            }

            justBeforeEach {
                dogsRepository.getDogs(offset: 0, limit: 5).pss_sink { _ in }.store(in: &cancellables)
                viewModel.onViewAppear()
                TimeAdvancingFactory(container).tick()
            }

            Then("The favorites are empty") {
                expect(viewModel.state) == .empty
            }

            When("I favorite two dogs") {
                justBeforeEach {
                    let dogs = dogsRepository.dogsFeed.record().lastValue().dogs
                    favoritesRepository.addFavorite(dog: dogs[3]).pss_sink { _ in }.store(in: &cancellables)
                    favoritesRepository.addFavorite(dog: dogs[1]).pss_sink { _ in }.store(in: &cancellables)
                    TimeAdvancingFactory(container).tick()
                }

                Then("The favorites grid shows them in grid order") {
                    guard case .loaded(let cells) = viewModel.state else { return fail("Expected loaded state") }
                    expect(cells.map(\.name)) == ["Dog 2", "Dog 4"]
                }

                And("I remove one favorite") {
                    justBeforeEach {
                        let dogs = dogsRepository.dogsFeed.record().lastValue().dogs
                        favoritesRepository.removeFavorite(dog: dogs[1]).pss_sink { _ in }.store(in: &cancellables)
                        TimeAdvancingFactory(container).tick()
                    }

                    Then("Only the remaining favorite is shown") {
                        guard case .loaded(let cells) = viewModel.state else { return fail("Expected loaded state") }
                        expect(cells.map(\.name)) == ["Dog 4"]
                    }
                }
            }
        }
    }
}

File: WoofAndroid/presentation/favorites/src/test/kotlin/com/perrystreet/woof/presentation/favorites/viewmodel/FavoritesViewModelTest.kt
package com.perrystreet.woof.presentation.favorites.viewmodel

import com.perrystreet.woof.datasource.dogs.DogsDataSourceFactory
import com.perrystreet.woof.presentation.favorites.spec.ViewModelBehaviorSpec
import com.perrystreet.woof.presentation.favorites.utils.UiObservableExtensions.test
import com.perrystreet.woof.repositories.dogs.DogsRepository
import com.perrystreet.woof.repositories.favorites.FavoritesRepository
import com.perrystreet.woof.testutils.rx.lastValue
import com.perrystreet.woof.testutils.scheduler.TimeAdvancingFactory
import io.kotest.matchers.types.shouldBeInstanceOf
import org.amshove.kluent.shouldBeEqualTo
import org.koin.test.inject

class FavoritesViewModelTest : ViewModelBehaviorSpec() {
    private val viewModel: FavoritesViewModel by inject()
    private val dogsRepository: DogsRepository by inject()
    private val favoritesRepository: FavoritesRepository by inject()

    init {
        Given("I open the favorites tab with dogs loaded in the grid") {
            TimeAdvancingFactory().withTestSchedulers()
            DogsDataSourceFactory().withDogs(count = 5)

            beforeEach {
                dogsRepository.getDogs(offset = 0, limit = 5).test()
                viewModel.onViewAppear()
                TimeAdvancingFactory().tick()
            }

            Then("The favorites are empty") {
                viewModel.state.test().lastValue() shouldBeEqualTo FavoritesViewModel.State.Empty
            }

            When("I favorite two dogs") {
                beforeEach {
                    val dogs = dogsRepository.dogsFeed.test().lastValue().dogs
                    favoritesRepository.addFavorite(dogs[3])
                    favoritesRepository.addFavorite(dogs[1])
                    TimeAdvancingFactory().tick()
                }

                Then("The favorites grid shows them in grid order") {
                    val state = viewModel.state.test().lastValue().shouldBeInstanceOf<FavoritesViewModel.State.Loaded>()
                    state.cells.map { it.name } shouldBeEqualTo listOf("Dog 2", "Dog 4")
                }

                And("I remove one favorite") {
                    beforeEach {
                        val dogs = dogsRepository.dogsFeed.test().lastValue().dogs
                        favoritesRepository.removeFavorite(dogs[1])
                        TimeAdvancingFactory().tick()
                    }

                    Then("Only the remaining favorite is shown") {
                        val state = viewModel.state.test().lastValue().shouldBeInstanceOf<FavoritesViewModel.State.Loaded>()
                        state.cells.map { it.name } shouldBeEqualTo listOf("Dog 4")
                    }
                }
            }
        }
    }
}
